File: packages/contracts/src/models/subtitle.ts
import { z } from "zod";
import { assetIdSchema, subtitleSessionIdSchema } from "./ids.js";

export const subtitleTransportSchema = z.literal("stdio");

export const subtitlePingSchema = z.object({
  service: z.literal("subtitle-sidecar"),
  protocolVersion: z.string().min(1),
  transport: subtitleTransportSchema,
});

export const subtitleHealthSchema = subtitlePingSchema.extend({
  nodeVersion: z.string().min(1),
  sharpVersion: z.string().min(1).optional(),
  uptimeMs: z.number().int().nonnegative(),
  activeSessions: z.number().int().nonnegative(),
  sessionsRoot: z.string().min(1),
});

export const subtitleHostSchema = z.object({
  executable: z.string().min(1),
  entryPath: z.string().min(1),
  running: z.boolean(),
  restartCount: z.number().int().nonnegative(),
  lastError: z.string().min(1).optional(),
  ping: subtitlePingSchema.optional(),
  health: subtitleHealthSchema.optional(),
});

export const subtitleSessionStateSchema = z.enum([
  "idle",
  "running",
  "stopped",
  "failed",
]);

export const subtitleSessionSchema = z.object({
  sessionId: subtitleSessionIdSchema,
  assetId: assetIdSchema.optional(),
  state: subtitleSessionStateSchema,
  progress: z.number().min(0).max(1),
  createdAt: z.string(),
  updatedAt: z.string(),
  outputPath: z.string().min(1).optional(),
});

export const subtitleProgressSchema = z.object({
  sessionId: subtitleSessionIdSchema,
  state: subtitleSessionStateSchema,
  progress: z.number().min(0).max(1),
  message: z.string().min(1).optional(),
});

export type SubtitleTransport = z.infer<typeof subtitleTransportSchema>;
export type SubtitlePing = z.infer<typeof subtitlePingSchema>;
export type SubtitleHealth = z.infer<typeof subtitleHealthSchema>;
export type SubtitleHost = z.infer<typeof subtitleHostSchema>;
export type SubtitleSessionState = z.infer<typeof subtitleSessionStateSchema>;
export type SubtitleSession = z.infer<typeof subtitleSessionSchema>;
export type SubtitleProgress = z.infer<typeof subtitleProgressSchema>;

File: crates/shared-model/src/subtitle.rs
use crate::{AssetId, SubtitleSessionId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitlePingSummary {
    pub service: String,
    pub protocol_version: String,
    pub transport: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleHealthSummary {
    pub service: String,
    pub protocol_version: String,
    pub transport: String,
    pub node_version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sharp_version: Option<String>,
    pub uptime_ms: u64,
    pub active_sessions: u64,
    pub sessions_root: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleHostSummary {
    pub executable: String,
    pub entry_path: String,
    pub running: bool,
    pub restart_count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ping: Option<SubtitlePingSummary>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub health: Option<SubtitleHealthSummary>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SubtitleSessionState {
    Idle,
    Running,
    Stopped,
    Failed,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleSessionSummary {
    pub session_id: SubtitleSessionId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,
    pub state: SubtitleSessionState,
    pub progress: f64,
    pub created_at: String,
    pub updated_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_path: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleProgressEvent {
    pub session_id: SubtitleSessionId,
    pub state: SubtitleSessionState,
    pub progress: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}
